### PR TITLE
SOL-150813: mock SEMP backend PoC for scalability testing

### PR DESCRIPTION
**Pull request by helen-zhu-solace**
## Summary

Adds `test/performance/` — a self-contained perf-testing that lets MCP be load-tested against a fleet of 50 mock Solace brokers without provisioning real infrastructure. Ships four Go commands (mock-semp, loadgen, fidelity, memsampler), driver shell scripts, and MCP configs for single-host and split-host runs.

Fixes [SOL-150813 ](https://sol-jira.atlassian.net/browse/SOL-150813)

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] Performance improvement

## Motivation and Context

We had no way to measure MCP throughput, latency, or memory behavior under realistic multi-broker load. Standing up 50 real brokers per run is impractical, and a single-broker setup caps at ~3 tool/s because of MCP's per-broker semaphore and `request_min_interval` — every client queues on one broker.

## Changes Made

- **`mock-semp/`** — HTTP replayer that binds N localhost ports and returns canned SEMPv1/v2 responses per broker. Supports a per-port config endpoint (`POST /_mock/config`) for latency injection and weighted retryable-error injection (429/500/502/503/504). Latency knob is the primary lever for making MCP's per-broker semaphore queue observable.
- **`loadgen/`** — MCP client that opens one session per `-clients` unit (mirrors the one-session-per-agent pattern) and hammers configured tools for a fixed duration. Reports throughput, error rate, and latency percentiles.
- **`fidelity/`** — Hard gate before any load run. Deep-equals live MCP tool output against golden JSON captured against a real broker; non-empty diff aborts the performance test. 
- **`memsampler/`** — Polls `/proc/<pid>/status` and writes RSS/VmSize/threads to CSV so a load run can be plotted against memory to catch leaks (target: <10% drift over 30s).
- **Driver scripts** — `build.sh`, `run.sh` (single-host smoke), `run-mcp.sh` / `run-loadgen.sh` (split-host demo), `sampler.sh` / `loadgen-sampler.sh` / `summary.sh` (measurement wrappers), `regen-golden.sh` (fidelity golden refresh).
- **MCP configs** — `broker-config.mock.yaml` (50-alias fan-out over ports 18081–18130) and `broker-config.real.yaml` (baseline against a real broker for golden capture).
- **Golden fixtures** — `fidelity/golden/` and `mock-semp/canned/` capture reference SEMP responses.
- `README.md` outlines purpose, commands, variables for script and ports. 

## Testing

**Test Environment:**
- Go version: 1.25.0
- OS: Linux (dev2 + split-host demo boxes)
- Broker version (if applicable): real broker used only for golden capture

**Tests Performed:**
- [ ] Unit tests added/updated
- [ ] Integration tests added/updated (if applicable)
- [ ] E2E tests added/updated (if applicable)
- [x] Tested locally with `go test -race ./...`
- [x] Tested with real Solace broker (fidelity golden capture)

**Test Coverage:**
- Fidelity gate passes: MCP-against-mock output deep-equals MCP-against-real-broker golden for `get-broker-status` and `list-queues`.
- Demo bar met per plan step 7: <0.5% errors, over 60s.
- Memory bar met per plan step 8: <10% RSS drift over 30s at steady state.
- Backpressure exercised by dialing `LATENCY_MS` so the per-broker semaphore (`semp.max_concurrent_per_broker`) queues callers inside MCP.
- Retry-chain exercised by injecting weighted 429/500/503 error responses via `/_mock/config`.

## Breaking Changes

None. All additions are under `test/performance/` and are not built or imported by production code.

**Impact:** N/A

**Migration Guide:** N/A

### Code Quality
- [x] Code follows the coding standards
- [x] Self-review completed
- [x] Code is well-commented (explains "why", not "what")
- [x] Complex logic has explanatory comments
- [x] No commented-out code or debug statements left in

### Security
- [x] No credentials in code (mock accepts any non-empty basic-auth)
- [x] Followed secure logging rules
- [x] Credential-carrying types implement `slog.LogValuer` (N/A — no new credential types)
- [x] No security vulnerabilities introduced

### Testing
- [x] All tests pass locally: `go test -race ./...`
- [x] No race conditions detected
- [x] Edge cases covered (error injection, latency queueing, session churn)
- [x] Error paths tested (fidelity divergence, injected retryables)
- [x] Test names clearly describe what is being tested

### Documentation
- [x] README.md updated (not user-facing — internal perf tooling)
- [x] godoc comments added/updated for exported functions (package docs on each `main.go`)
- [ ] CHANGELOG.md updated (not applicable — test-only tooling)
- [x] Examples updated (usage blocks in each command's package doc + script headers)

### Git & CI
- [x] All commits are signed off (DCO)
- [x] Commits have clear, descriptive messages
- [x] Branch is up to date with main
- [x] No merge conflicts
- [x] CI checks passing

## Screenshots (if applicable)
Broker MCP Server 
image

Loadgen + Mock Broker
image

## Additional Notes

- This intentionally lives under `test/performance/` and is not wired into CI. It's a lab tool, not a regression harness.
---

**For Reviewers:**

**Focus Areas**:
- `mock-semp/handler.go` — request routing, latency/error injection semantics, per-port config isolation.
- `loadgen/main.go` — one-session-per-client assumption; latency percentile aggregation under high concurrency.
- `fidelity/main.go` — deep-equal diffing (paths reported as `path: golden=<x> actual=<y>`).
- Shell drivers — env-var contract (`CLIENTS`, `DURATION`, `LATENCY_MS`, `ERROR_RATE`, `ERROR_STATUSES`) should be stable enough for the demo runs.

### CODE REVIEW

File: test/performance/loadgen/main.go [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+				tools:    cfg.tools,
+				warmup:   cfg.warmup,
+				duration: cfg.duration,
+				rps:      cfg.rps,
+				start:    startBarrier,
+				calls:    &totalCalls,
+				errs:     &totalErrs,
+			})
+		})
+	}
```
[the remark this answers is not in the document]
Comment (helen-zhu-solace) [reply test/performance/loadgen/main.go:183]: Not valid -  Go 1.25 added sync.WaitGroup.Go(func()), and this repo builds on Go 1.25.12 (go build ./test/perf-poc/... passes cleanly). The related "closes over loop variable i" claim is also stale — Go 1.22 made loop vars per-iteration. Marking as resolved.

File: test/performance/loadgen/main.go [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+	explicit := splitCSV(csv)
+	switch {
+	case count > 0 && len(explicit) > 0:
+		return nil, errors.New("-brokers and -broker-count are mutually exclusive")
+	case count > 0:
+		out := make([]string, count)
+		for i := range count {
+			out[i] = fmt.Sprintf("%s-%02d", prefix, i+1)
+		}
+		return out, nil
```
[the remark this answers is not in the document]
Comment (helen-zhu-solace) [reply test/performance/loadgen/main.go:573]: Not valid - Range-over-int has been valid since Go 1.22. for i := range count compiles. Marking as resolved.

File: test/performance/broker-config.mock.yaml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+# MOCK_HOST is required (MCP hard-fails on unset ${VAR}).
+#   - Single-host smoke: MOCK_HOST=localhost.
+#   - Split-host demo:   MOCK_HOST=<Box A LAN IP>, MCP runs on Box B.
+#
+# The mock accepts any credentials (it doesn't validate auth), so
+# BROKER_USERNAME / BROKER_PASSWORD can be set to any non-empty value —
+# MCP still requires them to build the basic-auth header.
+
+listen_address: 0.0.0.0
+allow_remote_unauthenticated: true
```
[the remark this answers is not in the document]
Comment (helen-zhu-solace) [reply test/performance/broker-config.mock.yaml:23]: Pushed a new comment in [6c19813](https://github.com/SolaceProducts/solace-broker-mcp/pull/245/commits/6c19813b9f5aa1a2a094d368fae51524f51a0efa) that leaves comment that this file is for lab use only.

File: test/performance/loadgen/main.go [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+func pctIdx(n int, p float64) int {
+	if n == 0 {
+		return 0
+	}
+	idx := int(float64(n) * p)
+	if idx >= n {
+		idx = n - 1
+	}
+	return idx
+}
```
[the remark this answers is not in the document]
Comment (helen-zhu-solace) [reply test/performance/loadgen/main.go:506]: fixed in [7b3c645](https://github.com/SolaceProducts/solace-broker-mcp/pull/245/commits/7b3c645f7c35b9d854820c03034253c899735b5f)

File: test/performance/mock-semp/handler.go [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+// randIntN returns a uniformly distributed int in [0, n) using crypto/rand.
+// Panics if n <= 0, matching math/rand/v2.IntN.
+func randIntN(n int) int {
+	if n <= 0 {
+		panic("randIntN: n must be > 0")
+	}
+	v, _ := rand.Int(rand.Reader, big.NewInt(int64(n)))
+	return int(v.Int64())
+}
```
[the remark this answers is not in the document]
Comment (helen-zhu-solace) [reply test/performance/mock-semp/handler.go:60]: fixed in [7b3c645](https://github.com/SolaceProducts/solace-broker-mcp/pull/245/commits/7b3c645f7c35b9d854820c03034253c899735b5f)

**Review by bczoma — APPROVED** [review]
Approved, awesome work, thank you!

File: test/performance/mock-semp/config.go [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+	502: true, // one retry
+	503: true, // full backoff chain, honors Retry-After — realistic "broker busy"
+	504: true, // one retry
+}
+
+// handler returns the HTTP handler for POST /_mock/config. GET returns
+// 405 — the endpoint is write-only, and a GET response would need to
+// snapshot every port's config, which is more machinery than this
+// suite needs.
+func (s *configStore) handler() http.Handler {
```
Comment (solace-aross) [inline test/performance/mock-semp/config.go:127]: `POST /_mock/config` isn't atomic: validation and `s.set()` happen in the same loop over `req.Ports` (a Go map, so unordered), and `s.set` does a live, unconditional `p.Store(&o)` with no staging or rollback. If any one port in a batch fails validation (bad `error_rate`, unknown port, etc.), the handler returns 400 for the whole request — but every port whose turn in map iteration came before the invalid one keeps its new override live. The caller has no way to tell from the 400 that anything changed.

Confirmed by injection (throwaway `httptest` test against a 50-port store, one port deliberately invalid): 400 returned, and anywhere from ~3 to 46 of the other 50 ports had already been mutated depending on map-iteration order — never zero. An Opus review pass reproduced this independently with its own harness and the same conclusion.

Not reachable today: both `run.sh` and `run-loadgen.sh` send identical values for every port in a batch, so a bad value is bad on the first port touched and 400s before anything is mutated (0/50 in that shape). But the endpoint's own doc comment (`config.go:92`) calls a *differentiated* per-port request ("add latency to broker 3") the common case, and `README.md` tells `NO_MOCK=1` operators to POST this by hand — so a hand-crafted differentiated request hits it on the first try.

Given this is unwired lab tooling with zero test coverage today and no production or CI exposure, I'm not blocking on it — but it's a real bug in the file the description flagged as a focus area, and it's a small fix: split into two passes, validate every entry into a `map[int]portOverride` first, then only call `s.set` once all of them pass.

**Review by solace-aross — APPROVED** [review]
Reviewed the two focus areas from the description (`mock-semp/handler.go`, `config.go`) plus the Makefile/coverage-doc change, which is the one piece of this PR that touches shared repo tooling rather than the lab harness itself.

Found one real bug, verified by injection, not blocking (details on the line). Everything else checks out:

- `make check`-equivalent: `go build ./...` clean, `go vet` clean at this head.
- The coverage-gate carve-out (`Makefile`, `docs/internal/unit-test-coverage.md`) does what it says: `go tool cover -func` on this head gives 88.5% with `test/` filtered vs 73.5% unfiltered, so the ~2,150 new statements would otherwise have forced the floor down or forced tests onto throwaway tooling. Checked the one thing worth checking on a policy change like this — whether the filter is scoped to what it claims — and it is in effect right now: `test/integration` is the only other package under top-level `test/` in `./...`, and it profiles as `[no statements]`, so today the broader `^test/` prefix and a narrower `test/performance/`-only prefix produce identical numbers (88.5% either way). Worth knowing for later: the exclusion is anchored on the whole `test/` prefix, not `test/performance/` specifically, so any future top-level `test/` Go package would silently join the exemption with no discussion required. Not asking for a change — the two constraints already written into the doc (categorical prefix, module-anchored so `internal/`/`cmd/` can't match) are the right guardrails — just flagging it as the thing to watch if this file changes again.
- Copilot's still-open thread on `regen-golden.sh:123` is a false positive — I pulled `broker_url_from_config()` out and ran it directly against double-quoted, single-quoted, and `${VAR}`-style URLs. All three resolve correctly; the close-quote/backslash-quote/reopen-quote sequence around each embedded `'` is the standard escape for a literal single quote inside a single-quoted bash string, not a break. Safe to resolve that thread.

Approving.

File: test/performance/mock-semp/config.go (hunk 6 above)
Comment (helen-zhu-solace) [reply test/performance/mock-semp/config.go:127]: Will open a follow up ticket that mentions this issue.
